kernel/arch/riscv/paging.c

```diff
@@ -163,9 +176,88 @@ void map_big_page_int(pdir_t *pdir,
    NOT_IMPLEMENTED();
 }
 
+/*
+ * Used in early boot(before kmalloc initialization),
+ * only big pages are mapped.
+ */
+static inline void
+create_early_page_table(pdir_t *pdir,
+                        ulong vaddr,
+                        ulong paddr,
+                        size_t lenth,
+                        bool in_vm)
+{
+   ulong end_va = vaddr + lenth;
+   ulong pmd_paddr;
+   page_t *e;
+   page_table_t *tmpdir;
+
+   for (; vaddr < end_va;) {
+
+      tmpdir = pdir;
+
+      for (int level = RV_PAGE_LEVEL; level > 0; level--) {
+
+         e = &tmpdir->entries[PTE_INDEX(level, vaddr)];
+
+         if ((level > 1) && (!e->present)) {
+
+            if (in_vm)
+               pmd_paddr = KERNEL_VA_TO_PA((ulong)alloc_early_pt());
+            else
+               pmd_paddr = (ulong)alloc_early_pt();
+
+            e->raw = (PFN(pmd_paddr) << _PAGE_PFN_SHIFT) | PAGE_TABLE;
+
+            tmpdir = (page_table_t *)pmd_paddr;
+         } else {
+
+            tmpdir = (void *)((ulong)e->pfn << PAGE_SHIFT);
+         }
+      }
+
+      e->raw = MAKE_BIG_PAGE(paddr);
+
+      vaddr += L1_PAGE_SIZE;
+      paddr += L1_PAGE_SIZE;
+   }
+}
+
 void *init_early_mapping(ulong fdt_paddr)
 {
-   NOT_IMPLEMENTED();
+   extern char _start;
+
+   pdir_t *pgd = (void *)&page_size_buf[0];
+   ulong kernel_pa = (ulong)&_start - 0x1000;
+   ulong text_offset = ((struct linux_image_h *)(void *)&_start)->text_offset;
+   ulong base_pa = kernel_pa - text_offset;
+
+   kernel_va_pa_offset = KERNEL_BASE_VA - base_pa;
+   linear_va_pa_offset = BASE_VA - base_pa;
+
+   /* Copy the flattened device tree to 1MB below kernel's head */
+   memcpy((void *)(kernel_pa - MB), (void *)fdt_paddr, MB);
+
+   /* Kernel physical address must be aligned to L1 level big pages(2/4 MB) */
+   ASSERT(!((ulong)kernel_pa & (L1_PAGE_SIZE - 1)));
+
+   /* Identity map the first 8 MB */
+   create_early_page_table(pgd, base_pa, base_pa, EARLY_MAP_SIZE, false);
+
+   /* Map the first 8 MB also at BASE_VA */
+   create_early_page_table(pgd, BASE_VA, base_pa, EARLY_MAP_SIZE, false);
+
+   if (!KRN32_LIN_VADDR) {
+      /* Map the first 8 MB AT KERNEL_BASE_VA*/
+      create_early_page_table(pgd, KERNEL_BASE_VA, base_pa,
```

Comment:
Style: it's nicer to have one argument per line for such a big calls.

Reply:
fixed!